### PR TITLE
Add maximize parameter to LBFGS optimizer

### LINKED ISSUE #126642: [optim] add missing 'maximize' parameter to LBFGS, NAdam and RAdam optimizers
### 🚀 The feature, motivation and pitch

Currently, ```LBFGS```, ```NAdam``` and ```RAdam``` optimizers do not have the ```maximize``` parameter in their constructors, which means they cannot be used with some loss functions (like R2Score).

### Alternatives

In the case ```maximize``` can't be implemented due to a constraint of the optimizer, it would be nice to make it clear that ```maximize``` is intentionally not supported. Some possible suggestions:
* Update the documentation to mention the reason maximize isn't supported for the optimizer
* Still support the parameter but throw an error when maximize=True

### Additional context

_No response_

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

### PR DESCRIPTION
**Pull request by rajfirke**
Fixes #126642

LBFGS is the only standard PyTorch optimizer without `maximize` support. A prior attempt (PR #81098) was closed because it only negated the gradient in one code path, causing the line search to fail.

Unlike simpler optimizers (SGD, Adam) where negating the gradient is sufficient, LBFGS also requires negated loss values because its strong Wolfe line search uses both `f(x)` and `∇f(x)` — both must be consistent for the Armijo and curvature conditions to evaluate correctly on the transformed objective `-f(x)`.

**Implementation:**
- Negate gradients inside `_gather_flat_grad()` when `maximize=True`, covering all three call sites (initial eval, line search via `_directional_evaluate`, and fixed-step re-eval)
- Negate loss in `_directional_evaluate` and at each `closure()` eval point in `step()`, so the line search operates on `-f(x)`
- Return original `orig_loss` to the caller (not negated)

**Changes:**
- `torch/optim/lbfgs.py`: Add `maximize` keyword-only parameter, negate grad+loss at all eval points
- `torch/testing/_internal/common_optimizers.py`: Add `maximize` and `maximize_strong_wolfe` test inputs

**Why this works mathematically:** Maximizing `f(x)` is equivalent to minimizing `g(x) = -f(x)`. By providing `∇g = -∇f` and `g(x) = -f(x)` to the L-BFGS algorithm, the two-loop recursion, H_diag scaling, and Wolfe line search all operate correctly on `g` and converge to a maximum of `f`.

cc @janeyx99 @albanD @jbschlosser

### CODE REVIEW

**Comment by rajfirke** [comment]
Friendly ping — adds `maximize` to LBFGS, the last optimizer without it. Negates both gradient and loss for correct Wolfe line search behavior. cc @janeyx99 @zou3519

**Comment by rajfirke** [comment]
Friendly ping — LBFGS maximize: negates both gradient and loss for correct Wolfe conditions. cc @janeyx99 @zou3519

**Comment by rajfirke** [comment]
Gentle ping — adds `maximize` parameter to LBFGS, the only optimizer missing it. Negates both gradient and loss for correct Wolfe conditions. cc @janeyx99 @zou3519

**Comment by janeyx99** [comment]
@claude review please using the skilll

**Comment by janeyx99** [comment]
As Claude does its review, I have some comments:
- approach looks good to me
- please add docs just like with all other optims

**Comment by rajfirke** [comment]
Thank you @janeyx99 for the review and @huydhn for the rebase!

Added the `maximize` docstring to the Args section, matching the same wording used in Adam, SGD, and other optimizers. Let me know if anything else is needed.

File: torch/optim/lbfgs.py [hunk 1]
```diff
@@ -235,6 +235,8 @@ class LBFGS(Optimizer):
             value/parameter changes (default: 1e-9).
         history_size (int, optional): update history size (default: 100).
         line_search_fn (str, optional): either 'strong_wolfe' or None (default: None).
+        maximize (bool, optional): maximize the objective with respect to the
```
Comment (janeyx99) [inline torch/optim/lbfgs.py]: Actually can you reuse https://github.com/pytorch/pytorch/blob/main/torch/optim/optimizer.py#L285C1-L285C14 like all other optims?

Comment (rajfirke) [reply torch/optim/lbfgs.py]: Done — switched to `LBFGS.__doc__` assignment with `{_maximize_doc}` from `optimizer.py`, matching the pattern used by SGD, Adam, and all other optimizers.

**Comment by rajfirke** [comment]
Hi @janeyx99 — the `_maximize_doc` fix is pushed per your feedback. Looks like the merge bot needs a formal approval to proceed. Would you be able to approve when you get a chance? Thank you for the review!

**Review by janeyx99 — APPROVED** [review]
thank you! sorry this fell under the bed in terms of backlog
